### ISSUE TITLE
Is dqlite really a single-writer DB?

### ISSUE THREAD

**Issue by skunkiferous** [canonical/dqlite#151]
From the FAQ "How does dqlite behave during conflict situations?", it seems that all write operations have to be performed in the master node, and therefore the non-master nodes can only make read-queries. This would require that the "user code" on non-master nodes somehow record all desired changes, and use some other networking stack to send those changes to the master to be executed. This seems very complicated and impractical. Is this correct? Could this be more precisely explained in the FAQ?

**Comment by freeekanayaka** [comment]
What you say is accurate. It's complicated but not impractical, and it's the only way you can have a fault tolerant highly-available data store.

I'll update the FAQ.

**Comment by skunkiferous** [comment]
I believe, some other distributed stores would let you do insert/delete/... on any node, but simply forward the request to the master node transparently, instead of running it locally. That would be "practical"...

OTOH, idk how well this would work with the SQLite API, which is designed for embedded use.

My train of thought is, setting up a reliable network delivery is difficult. Since you already have one as part of your product, taking this burden off your users seems like a good idea, but I assume this is not needed for your current use-case, or it would already be part of the feature-set.

**Comment by socketpair** [comment]
Well, What will happen in this sequence for 3-node cluster ?

1. current master is managed to write to it's local WAL
2. (power outage)
3. System find out that old master is not available at the moment, but  two old relicas are.
4. Since two of three replicas are available -- quorum is established. Start working in degraded mode.
5. client makes some INSERT
6. the old master become available.
7. conflicting data from old master's WAL?

**Comment by socketpair** [comment]
#46 ?

**Comment by skunkiferous** [comment]
Hmm. I never *coded* anything like that; only read about it. I believe that once a *new* leader is voted, if the old leader comes back online, it's state has to be invalidated, and it must query the new leader to get back up-to-date. So when the client makes an INSERT and sends it, it is associated with the leader that the client knows at that point. If it is the old leader, it will never get executed, but the client needs to be told that, eventually. If the client knows the new master, then it should get executed there, and the old-master resurfacing should have no effect on that.

**Comment by socketpair** [comment]
1. When does exactly client receives information that operation is complete? it shouldl only after quorum of nodes has notified that the operation is fsynced() in their WALs. 
2. AFAIK it is impossible to rollback operation in local sqlite DB. So in order to synchronize, that old master must re-download the whole DB.

**Comment by freeekanayaka** [comment]
@skunkiferous fwiw I believe the only requirement which might not be met by dqlite is "using some mechanism to partition data, such that processing can be moved where the data is": in dqlite all nodes replicate all data, there's no data partitioning, sharding or map/reduce-like feature.

**Comment by freeekanayaka** [comment]
@socketpair what @skunkiferous said is basically accurate. The changes that the leader wanted to commit will be rolled back as soon as it comes back to life. Thinking in terms of how SQLite works can be a bit misleading.

**Comment by mwild1** [comment]
Sorry about dragging up a closed issue, but I'm a bit confused after reading this discussion.

Is it absolutely correct if a non-leader *application* uses the dqlite API to commit a write transaction, it will just fail?

I'm asking because alternative projects, such as rqlite (to which dqlite is very often compared) will automatically send writes to the leader, the *application* itself doesn't have to be aware of this.

I don't even see an API in dqlite.h to let an application process discover whether it is currently the leader (or what node is currently the leader), so I'm not sure how this would be implemented at the application level anyway.

Maybe I'm just misunderstanding something.

**Comment by freeekanayaka** [comment]
> Sorry about dragging up a closed issue, but I'm a bit confused after reading this discussion.
> 
> Is it absolutely correct if a non-leader _application_ uses the dqlite API to commit a write transaction, it will just fail?

Did you read https://dqlite.io/docs/architecture ?

To put it shortly: if you are using `go-dqlite` (the only client we support and know of at the moment), everything will be taken care for you, any node of your application can run a SQL query and it will be automatically be run by the current leader.
 
> I'm asking because alternative projects, such as rqlite (to which dqlite is very often compared) will automatically send writes to the leader, the _application_ itself doesn't have to be aware of this.

With the `go-dqlite` client your application does not need to be aware of the current leader as well. Internal details  about how that's done may be different between dqlite and rqlite, but from the application point of view this particular aspect is the same.

> 
> I don't even see an API in dqlite.h to let an application process discover whether it is currently the leader (or what node is currently the leader), so I'm not sure how this would be implemented at the application level anyway.

The API to this sort of things is a wire-level API, not a C API. See https://dqlite.io/docs/protocol (the "Get current leader" request in particular.
 
> Maybe I'm just misunderstanding something.

As mentioned, the `go-dqlite` client is doing the heavy-lifting here, speaking the wire-level protocol and taking care of connecting to the current leader for running SQL queries. If you need support for a language other than Go, please let us know, there are plans for adding more clients.

**Comment by mwild1** [comment]
Ah, thanks for the clarification. So the additional work is handled by an additional layer between the application and the dqlite C code.

I had read the architecture page, however it doesn't mention Go, or C. The FAQ page says that the project is entirely in C.

Our application is in Lua, which has easy mechanisms to use C libraries. We already support sqlite, and I was expecting that dqlite was a replacement (with maybe some additional API for configuration). It seems right now that there is actually no full C client available. Specifically, I misunderstood this from statements in the FAQ such as:

> Embeddable in any language that can interoperate with C

and

> The project was later rewritten entirely in C
> [...]
> The added benefit of the rewrite in C is that it’s now easy to embed dqlite into project written in effectively any language since all major languages have provisions to create C bindings.

From what you're saying, it sounds like some of the critical logic to use dqlite seamlessly in a real application is actually in Go, and not C.

**Comment by cole-miller** [comment]
@mwild1 A C client library is absolutely on our roadmap, see #291.

**Comment by freeekanayaka** [comment]
> Ah, thanks for the clarification. So the additional work is handled by an additional layer between the application and the dqlite C code.
> 
> I had read the architecture page, however it doesn't mention Go, or C. The FAQ page says that the project is entirely in C.
> 
> Our application is in Lua, which has easy mechanisms to use C libraries. We already support sqlite, and I was expecting that dqlite was a replacement (with maybe some additional API for configuration). It seems right now that there is actually no full C client available. Specifically, I misunderstood this from statements in the FAQ such as:
> 
> > Embeddable in any language that can interoperate with C
> 
> and
> 
> > The project was later rewritten entirely in C
> > [...]
> > The added benefit of the rewrite in C is that it’s now easy to embed dqlite into project written in effectively any language since all major languages have provisions to create C bindings.
> 
> From what you're saying, it sounds like some of the critical logic to use dqlite seamlessly in a real application is actually in Go, and not C.

Right, all the information you read is accurate, but I see how it's prone to misunderstanding. I'd say it'd be nice to add some additional clarification to that documentation.

As @cole-miller a C client library is planned, at that point the client-side logic currently implemented in Go will also be available in C.
